core/trino-main/src/main/java/io/trino/cost/FilterStatsCalculator.java

```diff
@@ -169,35 +173,107 @@ protected PlanNodeStatsEstimate visitLogicalExpression(LogicalExpression node, V
+                    groupedExpressions.put(((ComparisonExpression) expression).getLeft(), expression);
+                    continue;
                 }
+                else if (expression instanceof InPredicate) {
+                    groupedExpressions.put(((InPredicate) expression).getValue(), expression);
+                    continue;
+                }
+                else if (expression instanceof NotExpression) {
+                    groupedExpressions.put(((NotExpression) expression).getValue(), expression);
+                    continue;
+                }
+                else if (expression instanceof IsNotNullPredicate) {
+                    groupedExpressions.put(((IsNotNullPredicate) expression).getValue(), expression);
+                    continue;
+                }
+                else if (expression instanceof IsNullPredicate) {
+                    groupedExpressions.put(((IsNullPredicate) expression).getValue(), expression);
+                    continue;
+                }
+                estimatesBuilder.add(process(expression));
             }
 
-            // If some of the filters cannot be estimated, take the smallest estimate.
-            // Apply 0.9 filter factor as "unknown filter" factor.
-            Optional<PlanNodeStatsEstimate> smallest = terms.stream()
-                    .map(this::process)
-                    .filter(termEstimate -> !termEstimate.isOutputRowCountUnknown())
-                    .sorted(Comparator.comparingDouble(PlanNodeStatsEstimate::getOutputRowCount))
-                    .findFirst();
+            for (Expression expression : groupedExpressions.keySet()) {
+                // calculate a combined estimate for each group so that we don't end up
+                // picking estimate from just one of the predicates in a BETWEEN clause
+                // or multiple range in-equalities on a symbol
+                List<Expression> expressions = groupedExpressions.get(expression);
+                PlanNodeStatsEstimate estimate = process(expressions.get(0));
+                if (!estimate.isOutputRowCountUnknown()) {
+                    for (int i = 1; i < expressions.size(); i++) {
+                        estimate = new FilterExpressionStatsCalculatingVisitor(estimate, session, types).process(expressions.get(i));
 
-            if (smallest.isEmpty()) {
-                return PlanNodeStatsEstimate.unknown();
+                        if (estimate.isOutputRowCountUnknown()) {
+                            break;
+                        }
+                    }
+                }
+                estimatesBuilder.add(estimate);
             }
+            return estimatesBuilder.build();
+        }
 
-            return smallest.get().mapOutputRowCount(rowCount -> rowCount * UNKNOWN_FILTER_COEFFICIENT);
+        private PlanNodeStatsEstimate combineEstimatesAnd(PlanNodeStatsEstimate combinedEstimate, PlanNodeStatsEstimate termEstimate, double outputRowCount, TypeProvider types)
```

Comment:
This is form of correlation. I think this should be called `intersectCorrelatedStats` and land in `PlanNodeStatsEstimateMath` (you can normalize afterwards).

It's still not perfect in terms of set operation definition, but IMO we can live with it